### PR TITLE
Improve failing log output of tests

### PR DESCRIPTION
**Pull request by norio-nomura**
On debugging failing tests, I did adhoc improvements sometime. But now, I think it should be implemented that properly.
## 

This depends on #832 
- Assert on each correction of `testSwiftLintAutoCorrects()`
- Produce detailed log of generated violations from `nonTriggeringExamples` : 

> /Users/norio/github/SwiftLint/Tests/SwiftLintFramework/TestHelpers.swift:136: error: -[SwiftLintFrameworkTests.RulesTests testMissingDocs] : failed - nonTriggeringExample violated:
> 
> ```
> /// docs
> public func a() {}
>        ^ warning: Missing Docs Violation: Public declarations should be documented. (missing_docs)
> ```
- Produce detailed log when triggeringExample without expected location does not violate:

> /Users/norio/github/SwiftLint/Tests/SwiftLintFramework/TestHelpers.swift:159: error: -[SwiftLintFrameworkTests.RulesTests testValidDocs] : failed - triggeringExample did not violate:
> 
> ```
> /// docs
> public func no() -> (Int, Int) {return (1, 2)}
> ```
- Produce detailed log when triggeringExample violated at unexpected location:

> /Users/norio/github/SwiftLint/Tests/SwiftLintFramework/TestHelpers.swift:173: error: -[SwiftLintFrameworkTests.RulesTests testMissingDocs] : failed - triggeringExample violate at unexpected location:
> 
> ```
> public func a↓() {}
>        ^ warning: Missing Docs Violation: Public declarations should be documented. (missing_docs)
> ```
- Produce detailed log when expected location of triggeringExample is not violated at:

> /Users/norio/github/SwiftLint/Tests/SwiftLintFramework/TestHelpers.swift:180: error: -[SwiftLintFrameworkTests.RulesTests testMissingDocs] : failed - triggeringExample did not violate at expected location:
> 
> ```
> public func a↓() {}
> ```
- Remove surplus assertions from triggeringExamples:
  - Comparing total count of violations and expected locations from all triggering examples
  - Comparing count of violations and expected locations on each triggering examples
  - Order based comparison between violations and expected locations

### CODE REVIEW

**Comment by jpsim** [comment]
This is unfortunately quite out of date due to depending on #832, and my lack of reviewing that.

@norio-nomura these seem like great improvements, maybe you could update this branch once #832 lands?

**Comment by norio-nomura** [comment]
Rebased.

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift [hunk 1]
```diff
@@ -34,6 +34,38 @@ private func cleanedContentsAndMarkerOffsets(from contents: String) -> (String,
     return (contents as String, markerOffsets.sorted())
 }
 
+func render(violations: [StyleViolation], in contents: String) -> String {
```
Comment (jpsim) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: `private`?

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift [hunk 2]
```diff
@@ -34,6 +34,38 @@ private func cleanedContentsAndMarkerOffsets(from contents: String) -> (String,
...
+        if line >= contents.count {
+            contents.append(message)
+        } else {
+            contents.insert(message, at: line)
+        }
+    }
+    return (["```"] + contents + ["```"]).joined(separator: "\n")
+}
+
+func render(locations: [Location], in contents: String) -> String {
```
Comment (jpsim) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: `private`?

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift [hunk 3]
```diff
@@ -34,6 +34,38 @@ private func cleanedContentsAndMarkerOffsets(from contents: String) -> (String,
...
+            "\(violation.ruleDescription.name) Violation: ",
+            violation.reason,
+            " (\(violation.ruleDescription.identifier))"].joined()
+        if line >= contents.count {
+            contents.append(message)
+        } else {
+            contents.insert(message, at: line)
+        }
+    }
+    return (["```"] + contents + ["```"]).joined(separator: "\n")
```
Comment (jpsim) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: Isn't this equivalent and simpler?

```swift
return "```\n\(contents)\n```"
```

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift [hunk 4]
```diff
@@ -34,6 +34,38 @@ private func cleanedContentsAndMarkerOffsets(from contents: String) -> (String,
...
+func render(locations: [Location], in contents: String) -> String {
+    var contents = (contents as NSString).lines().map { $0.content }
+    for location in locations.sorted(by: > ) {
+        guard let line = location.line, let character = location.character else { continue }
+        var content = contents[line - 1]
+        let index = content.index(content.startIndex, offsetBy: character - 1)
+        content.insert("↓", at: index)
+        contents[line - 1] = content
+    }
+    return (["```"] + contents + ["```"]).joined(separator: "\n")
```
Comment (jpsim) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: Isn't this equivalent and simpler?

```swift
return "```\n\(contents)\n```"
```

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift (hunk 3 above)
Comment (norio-nomura) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: `contents` is `[String]`.

File: Tests/SwiftLintFrameworkTests/TestHelpers.swift (hunk 4 above)
Comment (norio-nomura) [inline Tests/SwiftLintFrameworkTests/TestHelpers.swift]: Here also `contents` is [String]

**Pull request merged**